src/data/nsfg.clj

```diff
@@ -1,3 +1,165 @@
 (ns data.nsfg
   ^{:doc "Translation of python code to generate dataset used in ThinkStats2 book. Original source here: https://raw.githubusercontent.com/AllenDowney/ThinkStats2/master/code/nsfg.py"}
-  )
+  (:require [clojure.java.io :as io]
+            [clojure.string :as str]
+            [tablecloth.api :as tc])
+  (:import java.util.zip.GZIPInputStream))
+
+(def dict-line-regex #"^\s+_column\((\d+)\)\s+(\S+)\s+(\S+)\s+%(\d+)(\S)\s+\"([^\"]+)\"")
+
+(defn parse-dict-line
+  [line]
+  (let [[_ col type name f-len f-spec descr] (re-find dict-line-regex line)]
+    {:col    (dec (Integer/parseInt col))
+     :type   type
+     :name   (str/replace name "_" "-")
+     :f-len  (Integer/parseInt f-len)
+     :f-spec f-spec
+     :descr  descr}))
+
+(defn read-dictionary
+  "Read a Stata dictionary file, return a vector of column definitions."
+  [path]
+  (with-open [r (io/reader path)]
+    (mapv parse-dict-line (butlast (rest (line-seq r))))))
+
+(defn parse-value
+  [type raw-value]
+  (when (seq raw-value)
+    (case type
+      ("str12")          raw-value
+      ("byte" "int")     (Long/parseLong raw-value)
+      ("float" "double") (Double/parseDouble raw-value))))
+
+(defn make-row-parser
+  "Parse a row from a Stata data file according to the specification in `dict`.
+   Return a vector of columns."
+  [dict]
+  (fn [row]
+    (reduce (fn [accum {:keys [col type f-len]}]
+              (let [raw-value (str/trim (subs row col (+ col f-len)))]
+                (conj accum (parse-value type raw-value))))
+            []
+            dict)))
+
+(defn reader
+  "Open path with io/reader; coerce to a GZIPInputStream if suffix is .gz"
+  [path]
+  (if (.endsWith path ".gz")
+    (io/reader (GZIPInputStream. (io/input-stream path)))
+    (io/reader path)))
+
+(defn read-dictionary-data
+  "Parse lines from `rdr` according to the specification in `dict`.
+   Return a lazy sequence of parsed rows."
+  [dict rdr]
+  (let [parse-fn (make-row-parser dict)]
+    (map parse-fn (line-seq rdr))))
+
+(defn in?
+  "true if coll contains elm"
+  [coll elm]
+  (some #(= elm %) coll))
+
+(defn clean-fem-preg
+  " Recodes variables from the pregnancy frame.
+
+    # mother's age is encoded in centiyears; convert to years
+    df.agepreg /= 100.0
+
+    # birthwgt_lb contains at least one bogus value (51 lbs)
+    # replace with NaN
+    df.loc[df.birthwgt_lb1 > 20, 'birthwgt_lb1'] = np.nan
+
+    # replace 'not ascertained', 'refused', 'don't know' with NaN
+    na_vals = [97, 98, 99]
+    df.birthwgt_lb1.replace(na_vals, np.nan, inplace=True)
+    df.birthwgt_oz1.replace(na_vals, np.nan, inplace=True)
+
+    # birthweight is stored in two columns, lbs and oz.
+    # convert to a single column in lb
+    # NOTE: creating a new column requires dictionary syntax,
+    # not attribute assignment (like df.totalwgt_lb)
+    df['totalwgt_lb'] = df.birthwgt_lb1 + df.birthwgt_oz1 / 16.0"
+  [ds]
+  (-> ds
+      (tc/map-rows (fn [{:keys [agepreg]}]
```

Comment:
`map-columns` is supposed to be more efficient (and a bit less verbose).

It would be nice to use it in such cases.

Reply:
Good to know that 👍